mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp

```diff
@@ -0,0 +1,1638 @@
+
+      path.push_back(ifthenVal);
+
+      // TODO: bronching!
+
+      return codeGen(padIntermediateOp.array(), path, rewriter);
+    } else if (InOp inOp = dyn_cast<InOp>(val.getDefiningOp())) {
+      emitRemark(val.getLoc())
+          << "Codegen for In, generating read operation for operand";
+      return generateWriteAccess(path, inOp.input(), rewriter);
+    } else if (CastOp castOp = dyn_cast<CastOp>(val.getDefiningOp())) {
+      emitRemark(val.getLoc()) << "Codegen for Cast, reversing";
+      return generateWriteAccess(path, castOp.getOperand(), rewriter);
+    } else {
+      emitRemark(val.getLoc())
+          << "I don't know how to do codegen for: "
+          << val.getDefiningOp()->getName().getStringRef().str()
+          << " this is prob. an operation from another dialect. We walk "
+             "recursively through the operands until we hit something we can "
+             "do codegen for.";
+      int i = 0;
+      for (auto operand : (val.getDefiningOp()->getOperands())) {
+        val.getDefiningOp()->setOperand(i, codeGen(operand, path, rewriter));
+        i++;
+      }
+      return val;
+      // go through all the operands until we hit an idx
+    }
+  } else {
+    // val is a BlockArg
+    emitRemark(val.getLoc()) << "reached a blockArg in Codegen, reversing";
+    return generateWriteAccess(path, val, rewriter);
+  }
+}
+
+Value mlir::rise::generateWriteAccess(SmallVector<OutputPathType, 10> path,
+                                      Value accessVal,
+                                      PatternRewriter &rewriter) {
+  int index;
+  SmallVector<Value, 10> indexValues = {};
+  //  for (OutputPathType element : path) {
+  for (auto element = path.rbegin(); element != path.rend(); ++element) {
+    if (auto i = mpark::get_if<int>(&*element)) {
+      index = *i;
+    } else if (auto val = mpark::get_if<Value>(&*element)) {
+      indexValues.push_back(*val);
+    }
+  }
+  //  // handle problem originating from translation of reduce (accessing
+  //  element) int rank = accessVal.getType().dyn_cast<MemRefType>().getRank();
+  //  if (indexValues.size() != rank) {
+  //    indexValues.erase(indexValues.begin());
+  //  }
+  if (isa<AffineForOp>(rewriter.getBlock()->getParent()->getParentOp())) {
+    return rewriter
+        .create<AffineLoadOp>(accessVal.getLoc(), accessVal, indexValues)
+        .getResult();
+  } else {
+    return rewriter.create<LoadOp>(accessVal.getLoc(), accessVal, indexValues)
+        .getResult();
+  }
+}
+
+void mlir::rise::generateReadAccess(SmallVector<OutputPathType, 10> path,
+                                    Value storeVal, Value storeLoc,
+                                    PatternRewriter &rewriter) {
+  int index;
+  SmallVector<Value, 10> indexValues = {};
+  for (auto element = path.rbegin(); element != path.rend(); ++element) {
+    if (auto i = mpark::get_if<int>(&*element)) {
+      index = *i;
```

Comment:
Is it deliberate that `index` might be overridden multiple times?

Reply:
Index is actually never used here. This is a leftover from when it was not 100% clear what the path looks like. I removed that part